src/runtime/pythonengine.cs

```diff
@@ -64,7 +67,17 @@ public static string ProgramName
 
                 return result ?? "";
             }
-            set { Runtime.Py_SetProgramName(value); }
+            set
+            {
+                if (_programName != IntPtr.Zero)
+                {
+                    Marshal.FreeHGlobal(_programName);
+                }
+                _programName = Runtime.IsPython3
+                    ? UcsMarshaler.GetInstance("").MarshalManagedToNative(value)
```

Comment:
2 custom marshalling articles suggest that GetInstance should be a singleton like this:

```
public static ICustomMarshaler GetInstance(string cookie)
   {
       if (marshaler == null)
       {
              marshaler = new UcsMarshaler();
       }
       return marshaler;
   }
```